### PR TITLE
Faster assembly

### PR DESCRIPTION
**Pull request by atgeirr**
This is a draft showing work done to improve assembly times. Needs companion PR in opm-simulators.

All work done by copying and modifying existing classes, so as to avoid changing anything in the existing Flow, and to be able to compare.

### CODE REVIEW

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+                       unsigned timeIdx) const
+    {
+        // retrieve the source term intrinsic to the problem
+        elemCtx.problem().source(source, elemCtx, dofIdx, timeIdx);
+
+        // deal with MICP (if present)
+        MICPModule::addSource(source, elemCtx, dofIdx, timeIdx);
+
+        // scale the source term of the energy equation
+        if (enableEnergy)
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:453]: if constexpr

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                                 unsigned phaseIdx,
+                                 unsigned pvtRegionIdx,
+                                 const Eval& surfaceVolumeFlux,//value of RateVector
+                                 const FluidState& upFs)
+    {
+        //const auto& invB = getInvB_<FluidSystem, FluidState, UpEval>(upFs, phaseIdx, pvtRegionIdx);
+        //const auto& surfaceVolumeFlux = invB*extQuants.volumeFlux(phaseIdx);
+        unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::solventComponentIndex(phaseIdx));
+
+        if (blackoilConserveSurfaceVolume)
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:488]: if constexpr

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    enum { gasPhaseIdx = FluidSystem::gasPhaseIdx };
+    enum { oilPhaseIdx = FluidSystem::oilPhaseIdx };
+    enum { waterPhaseIdx = FluidSystem::waterPhaseIdx };
+
+    enum { gasCompIdx = FluidSystem::gasCompIdx };
+    enum { oilCompIdx = FluidSystem::oilCompIdx };
+    enum { waterCompIdx = FluidSystem::waterCompIdx };
+    enum { compositionSwitchIdx = Indices::compositionSwitchIdx };
+
+    static const bool waterEnabled = Indices::waterEnabled;
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:78]: constexpr these

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    static void adaptMassConservationQuantities_(Dune::FieldVector<Scalar, numEq>& container, unsigned pvtRegionIdx)
+    {
+        if (blackoilConserveSurfaceVolume)
+            return;
+
+        // convert "surface volume" to mass. this is complicated a bit by the fact that
+        // not all phases are necessarily enabled. (we here assume that if a fluid phase
+        // is disabled, its respective "main" component is not considered as well.)
+
+        if (waterEnabled) {
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:550]: if constexpr

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            unsigned activeWaterCompIdx = Indices::canonicalToActiveComponentIndex(waterCompIdx);
+            container[conti0EqIdx + activeWaterCompIdx] *=
+                FluidSystem::referenceDensity(waterPhaseIdx, pvtRegionIdx);
+        }
+
+        if (gasEnabled) {
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:556]: if constexpr

File: opm/models/blackoil/blackoillocalresidualtpfa.hh [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+            unsigned activeGasCompIdx = Indices::canonicalToActiveComponentIndex(gasCompIdx);
+            container[conti0EqIdx + activeGasCompIdx] *=
+                FluidSystem::referenceDensity(gasPhaseIdx, pvtRegionIdx);
+        }
+
+        if (oilEnabled) {
```
Comment (akva2) [inline opm/models/blackoil/blackoillocalresidualtpfa.hh:562]: if constexpr

**Review by akva2 — COMMENTED** [review]
Just some small things I noticed. This PR is not really mergeable as-is ,it needs some squashery and reorganization in logical bits.

**Comment by atgeirr** [comment]
> Just some small things I noticed. This PR is not really mergeable as-is ,it needs some squashery and reorganization in logical bits.

There is a lot of work to be done. I intended to open as a draft PR just clicked wrong.

**Pull request closed without merging**